src/main/java/nextstep/qna/service/QnAService.java

```diff
@@ -26,24 +26,10 @@ public class QnAService {
     @Transactional
     public void deleteQuestion(NsUser loginUser, long questionId) throws CannotDeleteException {
         Question question = questionRepository.findById(questionId).orElseThrow(NotFoundException::new);
-        if (!question.isOwner(loginUser)) {
-            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
-        }
-
-        List<Answer> answers = question.getAnswers();
-        for (Answer answer : answers) {
-            if (!answer.isOwner(loginUser)) {
-                throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
-            }
-        }
-
-        List<DeleteHistory> deleteHistories = new ArrayList<>();
-        question.setDeleted(true);
-        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, questionId, question.getWriter(), LocalDateTime.now()));
-        for (Answer answer : answers) {
-            answer.setDeleted(true);
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+
+        Question deleteQuestion = question.delete(loginUser);
+        DeleteHistories deleteHistories = deleteQuestion.addDeleteHistory();
```

Comment:
삭제를 수행하지 않고 addDeleteHistory 메서드를 호출하면 삭제 여부와 상관없이 삭제 이력 객체가 생성되는데요
실제로 삭제 되지 않는 내용을 저장하면 의도한대로 동작한다고 보기 어려울 것 같아요
삭제와 동시에 삭제 이력을 생성하면 어떨까요?
```suggestion
        DeleteHistories deleteHistories = question.delete(loginUser);
```

Reply:
위에 피드백 주신 내용도 그렇고 동의합니다.
> 삭제하지 않고 이 메서드를 호출하면 의도한대로 동작했다고 볼 수 있을까요?

삭제가 되었을때만 수행될 수 있도록 외부에 메서들 공개하지 않고, `delete` 호출 시 같이 수행되도록 수정하는게 좋을 것 같아요!

---

원래는 피드백 주신 구조처럼 `delete + addDeleteHistory` 두 로직을 하나의 메서드에서 처리했었는데요. ([commit ac8f59f](https://github.com/next-step/java-lms/pull/413/commits/ac8f59f171f754ba9176fc8bfa162485e5621598))

아래의 고민을 거쳐 결과적으로 현재의 코드로 수정하게 됐습니다.
* `delete` 메서드에서 `delete` 외에 `addHistory` 를 하고 있다.
* 그러면 메서드 명이 `delete` 가 아니라 `deleteAndAddHistory` 가 되는게 좋지 않을까?
* 메서드명이 2가지의 행위를 표현한다면 하나의 메서드가 여러 책임을 하고 있다고 볼 수 있을 것 같다.
* 그렇다면 분리해서 메서드의 재사용성을 높이는게 좋겠다.

근데 피드백을 통해 제가 놓쳤던 부분을 깨달았네요 🥹
> 삭제하지 않고 이 메서드를 호출하면 의도한대로 동작했다고 볼 수 있을까요?

> 실제로 삭제 되지 않는 내용을 저장하면 의도한대로 동작한다고 보기 어려울 것 같아요
삭제와 동시에 삭제 이력을 생성하면 어떨까요?

메서드의 재사용성을 높이는 것도 중요하지만, 
요구사항을 생각했을 때 외부에서 의도치 않게 호출되지 않도록, 적절한 호출 시점에 대한 고민이 선행되어야 할 것 같아요.

Reply:
[commit d5d382b](https://github.com/next-step/java-lms/pull/413/commits/d5d382b68b69479e48efb23675029564e55eaaa5) 으로 작업해서 반영했는데요.

`Commit suggestion` 에서 리뷰어님께서 제안해주신대로 바로 `commit` 하는 기능이 있네요!
다음번엔 해당 기능 사용해서 반영해볼게요 🙊